### PR TITLE
[CWS] TC classifiers & DNS event

### PR DESCRIPTION
**Pull request by Gui774ume**
### What does this PR do?

This PR introduces the first steps required to start working with BPF TC classifiers within CWS. TC classifiers are filters of the Linux Traffic Control subsystem that can be used to monitor and shape the network traffic on a host. TC classifiers are attached to qdiscs, which are themselves attached to network interfaces. A TC classifier can only see the ingress and egress traffic of its network interface, which means that we need to attach to all interfaces if we want to monitor the local traffic too. As such, this PR tracks the creation of new network interfaces so that we can dynamically add new TC classifiers on them. Since network interfaces are namespaced, we also needed to track network namespaces so that we can interact with them (and attach TC classifiers).

Another big part of this PR is to be able to route the ingress (or egress) traffic to the destination (or source) process that will receive (or that has emitted) the network traffic we intercept. This is done by monitoring the ports and IP addresses that are bound by the running processes on the host. A snapshot of these bindings was also added so that we can route the traffic of existing connections.

In between interfaces, various netfilter hook points can be used to redirect or mangle with network packets. Since we match processes with their binding ports and IP addresses, we also need to track any network address translation that might occur through the netfiler ABI. For now, we track the operations made only by the `nat` table.

In order to test these features, we implemented a kernel space DNS request parser and introduced the `dns` event. The following fields are exported:
- `dns.id`: the DNS request ID
- `dns.question.count`: the number of questions in the DNS request
- `dns.question.class`: the class of the DNS request
- `dns.question.type`: the type of the DNS request
- `dns.question.size`: the total size in bytes of the DNS request
- `dns.question.name`: the name of the DNS server

The network flow context of the DNS request can be accessed through the `network` top level keyword.

Last but not least, a new command was added to the `security-agent` binary: `security-agent runtime network-namespace dump` which will dump the list of network namespace held in cache. The `--snapshot-interfaces` can also be used to specify that a snapshot of the interfaces of each network-namespace should be performed during the dump.

Example rule:
```
---
version: 1.0.0
macros:
 - id: CRYPTOMINER_DOMAINS
   expression: |-
     [ "asia1.ethpool.org", "ca.minexmr.com", "cn.stratum.slushpool.com", "de.minexmr.com",
     "eth-ar.dwarfpool.com", "eth-asia.dwarfpool.com", "eth-asia1.nanopool.org", "eth-au.dwarfpool.com",
     "eth-au1.nanopool.org", "eth-br.dwarfpool.com", ... ]
     
rules:
 - id: potential_cryptominer
   description: A process resolved a DNS name associated to cryptomining activity
   expression: dns.question.name in CRYPTOMINER_DOMAINS
```

### Motivation

The goal of this PR is to start working with TC classifiers on a very simple example.

### Additional Notes

This PR is blocked on the following `ebpf-manager` PRs:
- https://github.com/DataDog/ebpf-manager/pull/22
- https://github.com/DataDog/ebpf-manager/pull/23
- https://github.com/DataDog/ebpf-manager/pull/25
- https://github.com/DataDog/ebpf-manager/pull/26

### Describe how to test/QA your changes

Functional tests were added to test the new `dns` event type.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by kayayarai — APPROVED** [review]

**Review by djmitche — COMMENTED** [review]
`pkg/config/config.go` change looks good for agent-core

File: pkg/security/probe/namespace_resolver.go [hunk 1]
```diff
@@ -0,0 +1,620 @@
...
+func (nn *NetworkNamespace) hasValidHandle() bool {
+	return nn.handle != nil
+}
+
+// NamespaceResolver is used to store namespace handles
+type NamespaceResolver struct {
+	sync.RWMutex
+	state     int64
+	probe     *Probe
+	resolvers *Resolvers
```
Comment (lebauce) [inline pkg/security/probe/namespace_resolver.go:197]: resolvers doesn't seem to be used

File: pkg/security/probe/namespace_resolver.go [hunk 2]
```diff
@@ -0,0 +1,620 @@
...
+
+// ResolveNetworkNamespace returns a file descriptor to the network namespace. WARNING: it is up to the caller to
+// close this file descriptor when it is done using it. Do not forget to close this file descriptor, otherwise we might
+// exhaust the host IPs by keeping all network namespaces alive.
+func (nr *NamespaceResolver) ResolveNetworkNamespace(nsID uint32) *NetworkNamespace {
+	if !nr.probe.config.NetworkEnabled || nsID == 0 {
+		return nil
+	}
+
+	nr.Lock()
```
Comment (lebauce) [inline pkg/security/probe/namespace_resolver.go:271]: RLock/RUnlock seems to be enough here

File: pkg/security/probe/namespace_resolver.go [hunk 3]
```diff
@@ -0,0 +1,620 @@
...
+			netnsDump.DevicesInQueue = append(netnsDump.DevicesInQueue, NetworkDeviceDump{
+				IfName:  dev.Name,
+				IfIndex: int(dev.IfIndex),
+			})
+		}
+
+		if params.GetSnapshotInterfaces() {
+			handle, err = netns.getNamespaceHandleDup()
+			if err != nil {
+				continue
```
Comment (lebauce) [inline pkg/security/probe/namespace_resolver.go:553]: netns.Unlock()

**Review by lebauce — CHANGES_REQUESTED** [review]

File: pkg/security/probe/namespace_resolver.go (hunk 3 above)
Comment (Gui774ume) [reply pkg/security/probe/namespace_resolver.go:553]: Nice catch 😱

File: pkg/security/probe/namespace_resolver.go (hunk 2 above)
Comment (Gui774ume) [reply pkg/security/probe/namespace_resolver.go:271]: 👍🏻

File: pkg/security/probe/namespace_resolver.go (hunk 1 above)
Comment (Gui774ume) [reply pkg/security/probe/namespace_resolver.go:197]: 👍🏻

**Review by lebauce — APPROVED** [review]

**Review by paulcacheux — APPROVED** [review]

**Pull request merged**